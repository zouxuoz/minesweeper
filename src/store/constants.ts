export enum GameStatus {
  PLAYING = 'PLAYING',
  WIN = 'WIN',
  LOSE = 'LOSE',
}

export const BOMB = 9;

export const START_GAME = 'START_GAME';
export const RESTART_GAME = 'RESTART_GAME';
export const RESET_GAME = 'RESET_GAME';
export const END_GAME = 'END_GAME';
export const REVEAL_CELL = 'REVEAL_CELL';
export const TOGGLE_FLAG = 'TOGGLE_FLAG';

export const SET_STATUS = 'SET_STATUS';
export const SET_CONFIG = 'SET_CONFIG';
export const SET_BOARD = 'SET_BOARD';
