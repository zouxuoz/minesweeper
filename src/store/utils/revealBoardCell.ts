import * as R from 'ramda';

import { BOMB } from '../constants';
import { BoardCell, GameState } from '../types';
import { revealBombs } from './revealBombs';
import { advancedFloodFill } from './advancedFloodFill';

export const revealBoardCell = (
  board: GameState['board'],
  columnIndex: number,
  rowIndex: number,
): GameState['board'] => {
  let nextBoard = board;

  const boardCell: BoardCell | undefined = R.path([rowIndex, columnIndex], board);

  if (!boardCell || boardCell.revealed || boardCell.flagged) {
    return nextBoard;
  }

  if (boardCell.value === BOMB) {
    nextBoard = R.assocPath([rowIndex, columnIndex, 'revealed'], true, nextBoard);
    nextBoard = revealBombs(nextBoard);
  } else if (boardCell.value === 0) {
    nextBoard = advancedFloodFill(
      nextBoard,
      columnIndex,
      rowIndex,
      R.assoc('revealed', true),
      cell => !cell.revealed && !cell.flagged && cell.value !== BOMB,
      cell => cell.value === 0 && !cell.revealed,
    );
  } else {
    nextBoard = R.assocPath([rowIndex, columnIndex, 'revealed'], true, nextBoard);
  }

  return nextBoard;
};
