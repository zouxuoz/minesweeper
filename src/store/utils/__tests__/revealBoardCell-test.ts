import { BOMB } from '../../constants';
import { revealBoardCell } from '../revealBoardCell';

const BOARD = [
  [
    { value: 0, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: BOMB, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
  ],
  [
    { value: 0, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
  ],
  [
    { value: 1, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: true },
  ],
  [
    { value: BOMB, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
  ],
  [
    { value: 1, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: 0, revealed: false, flagged: false },
    { value: 1, revealed: false, flagged: false },
    { value: BOMB, revealed: false, flagged: false },
  ],
];

it('Should reveal area by zero cell.', () => {
  expect(revealBoardCell(BOARD, 2, 2)).toMatchSnapshot();
});

it('Should reveal area by number cell.', () => {
  expect(revealBoardCell(BOARD, 2, 1)).toMatchSnapshot();
});

it('Should reveal all boms if cell contain bomb.', () => {
  expect(revealBoardCell(BOARD, 0, 3)).toMatchSnapshot();
});

it('Should return same board if cell is flagged', () => {
  expect(revealBoardCell(BOARD, 4, 2)).toEqual(BOARD);
});
