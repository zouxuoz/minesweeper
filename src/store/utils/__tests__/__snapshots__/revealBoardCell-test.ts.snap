// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should reveal all boms if cell contain bomb. 1`] = `
Array [
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": true,
      "revealed": false,
      "value": 0,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": true,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 9,
    },
  ],
]
`;

exports[`Should reveal area by number cell. 1`] = `
Array [
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": true,
      "revealed": false,
      "value": 0,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
  ],
]
`;

exports[`Should reveal area by zero cell. 1`] = `
Array [
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 0,
    },
    Object {
      "flagged": true,
      "revealed": false,
      "value": 0,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
  ],
  Array [
    Object {
      "flagged": false,
      "revealed": false,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 0,
    },
    Object {
      "flagged": false,
      "revealed": true,
      "value": 1,
    },
    Object {
      "flagged": false,
      "revealed": false,
      "value": 9,
    },
  ],
]
`;
